### ISSUE TITLE
Qute global values should probably be lazy

### ISSUE THREAD

**Issue by FroMage** [quarkusio/quarkus#39832]
### Describe the bug

While filing a LangChain4J issue related to rendering a prompt using Qute, which is probably using the wrong classloader, which invoked global values, I noticed that global values were not lazy: https://github.com/quarkiverse/quarkus-langchain4j/issues/436

So, there are two Qute issues here:

- Perhaps global values should be lazy, after all they're method calls?
- Probably global values should only be passed for certain targets? I mean, those were intended for HTML/Web templates, and while I can imagine them being useful even for email templates, I was surprised to see them evaluated for AI prompts. Perhaps we should add something of a target/scope to globals, like `Web, Email, AI`…?

**Comment by mkouba** [comment]
> Perhaps global values should be lazy,

Globals variables are implemented as `TemplateInstance.Initializer`s, see https://quarkus.io/guides/qute-reference#global_variables; i.e. they're added to the data map during initialization of a template instance. Which is exactly what I see in the stack trace.

> after all they're method calls?

...and fields ;-)

> Probably global values should only be passed for certain targets? I mean, those were intended for HTML/Web templates, and while I can imagine them being useful even for email templates, I was surprised to see them evaluated for AI prompts. Perhaps we should add something of a target/scope to globals, like Web, Email, AI…?

These variables are "global" by definition. If you only need something in a certain template, then don't use "global variable" but a custom `TemplateInstance.Initializer` instead :shrug:.

In theory, we could add something like a set of targets where the global is accessible. But this would require the creator of the `TemplateInstance` to add a special attribute with `TemplateInstance#setAttribute()`, i.e. something like:

```
@TemplateGlobal(targets = ["AI","web"])
public class Globals {
    static int age = 40;
}
```
and then when creating the template instance something like: 
```
template.instance().setAttribute("target", "AI").render();
```
so that we could compare `TemplateGlobal#targets` with `TemplateInstance#getAttribute("target")`.

And `target` is ofc not a good name at all.

**Comment by FroMage** [comment]
> they're added to the data map during initialization of a template instance.

Well, they could be added to the map as a "lazy cacheable" value.

`target` is not great, but we can find a better name later. Intuitively, I expect that we could default that `target` on `TemplateInstance` for anything declared in controllers as `web`? And `mail` for `MailTemplateInstance`?

I mean, both solutions would probably solve the issue of unwanted globals showing up: 

- lazy globals, because if you don't use them, they are not initialised
- segmented values depending on target, because AI prompts wouldn't see them

I'm not sure which is better, or if there's another even better approach? Both options do seem valuable for other use-cases, though, generally, no?

**Comment by mkouba** [comment]
> I'm not sure which is better, or if there's another even better approach? Both options do seem valuable for other use-cases, though, generally, no?

Yes, I think that we should make it possible to compute the data lazily and use it for global variables.
